### PR TITLE
refactor(graphs): Introduce `AreaMaskBuilder`

### PR DESCRIPTION
**Pull request by mpvginde**
When building a Stretched Grid or LAM hidden mesh from a refined Icosahedron, the mesh structure is defined by using a `KNNAreaMaskBuilder` for calculating which hidden nodes fall _under_ the high resolution part of the data mesh. 

This is currently done using a nearest neighbour search with `sklearn.neigbors.NearestNeighbors`. When going to very high refinement levels (>10), this becomes a bottleneck in the graph building (~1h)

This PR adds a new MaskBuilder-class: `RadiusAreaMaskBuilder` that uses a radius search either with `torch_geometric.nn.radius` when `torch-cluster` is installed or with `scipy.spatial.cKDTree` when `torch-cluster` is not available. 

This results in a 60% speedup. Some benchmarks:
| Builder | Precision | Where |Global Resolution | LAM resolution | MS-edge-resolution | Time |  
| --- | ---|--- |---| ---| ---|---|
|KNN          |`numpy.float64` |CPU |7 |10 |10 |4m 46s |
|~~DotProduct~~   |`torch.float64` |GPU |7 |10 |10 |3m 31s |
|~~DotProduct~~   |`torch.float32` |GPU |7 |10 |10 |2m 23s |
|pyg.nn.radius|`torch.float64` |CPU |7 |10 |10 |1m 24s |
|pyg.nn.radius|`torch.float64` |GPU |7 |10 |10 |1m 40s |
|KNN          |`numpy.float64` |CPU |8 |12 |12 |~ 1h   |
|~~DotProduct~~   |`torch.float32` |GPU |8 |12 |12 |~36m |
|pyg.nn.radius|`torch.float64` |CPU |8 |12 |12 | 14m 52s  |
|pyg.nn.radius|`torch.float64` |GPU |8 |12 |12 |17m 23s |
|scipy.spatial.cKDTree | `numpy.float?` | CPU |8 |12 |12 | 15m 1s |

Some figures confirming identical Stretched Grid Meshes (4,6) for the MEPS domain
image

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

### CODE REVIEW

File: graphs/src/anemoi/graphs/generate/masks.py [hunk 1]
```diff
@@ -9,15 +9,243 @@
...
+        * use_gpu = False: Calculations are done on the CPU (seems to be fasted from the benchmarks)
+    Else falling back to scipy.spatial.cKDTree (on CPU)
+    """
+
+    def __init__(
+        self,
+        reference_node_name: str,
+        margin_radius_km: float = 100,
+        mask_attr_name: str | None = None,
+        use_gpu: bool = False,
```
Comment (JPXKQX) [inline graphs/src/anemoi/graphs/generate/masks.py]: Looking at the results, shouldn't we use cpu always? It seems to be faster in both cases: with and without torch-cluster

Comment (mpvginde) [reply graphs/src/anemoi/graphs/generate/masks.py]: I was thinking maybe later we can exploit the fact that we have multiple GPU available and devide the work over the different GPUs for even more speed up. Not sure if what I say make sense. I don't know if `torch-cluster` supports parallelising this kind of functions/work?

Comment (JPXKQX) [reply graphs/src/anemoi/graphs/generate/masks.py]: As far as I know, only "single-device" parallelism is supported. We need something similar to the model sharding in anemoi-models (we should have the code to split into shards and gather the shards later). I think this will be easier to explore once we move the graph inside the model.

Comment (JPXKQX) [reply graphs/src/anemoi/graphs/generate/masks.py]: Even in that case, I would expect this to happen automatically, right?

**Comment by mpvginde** [comment]
Small update to the logging calls.
Logs now look like this:

**torch_cluster on CPU**
```bash
2026-05-22 15:23:07 DEBUG Initializing _TorchClusterAreaMaskBackend on device cpu
2026-05-22 15:23:08 INFO Fitting AreaMaskBuilder (_TorchClusterAreaMaskBackend) with 822681 reference nodes from data[cutout_mask].
2026-05-22 15:23:08 DEBUG Reference nodes live on cpu
2026-05-22 15:23:08 DEBUG _TorchClusterAreaMaskBackend: Moving reference vector coordinates from cpu to cpu
2026-05-22 15:23:18 INFO Computing area-mask (_TorchClusterAreaMaskBackend) for 2562 query nodes with margin: 11.0 km
2026-05-22 15:23:18 DEBUG Query nodes live on cpu
2026-05-22 15:23:18 DEBUG _TorchClusterAreaMaskBackend: Getting the query coordinate mask with Eucl. chord threshold 0.0017
2026-05-22 15:23:18 DEBUG _TorchClusterAreaMaskBackend: Moving query vector coordinates from cpu to cpu
2026-05-22 15:23:22 INFO Computing area-mask (_TorchClusterAreaMaskBackend) for 40962 query nodes with margin: 11.0 km
2026-05-22 15:23:22 DEBUG Query nodes live on cpu
2026-05-22 15:23:22 DEBUG _TorchClusterAreaMaskBackend: Getting the query coordinate mask with Eucl. chord threshold 0.0017
2026-05-22 15:23:22 DEBUG _TorchClusterAreaMaskBackend: Moving query vector coordinates from cpu to cpu
2026-05-22 15:23:23 DEBUG Time to register node coordinates (StretchedTriNodes): 15.97 s
2026-05-22 15:23:23 DEBUG Time to register node coordinates (StretchedTriNodes): 0.00 s
```

**torch_cluster on GPU**
```bash
2026-05-22 15:38:20 DEBUG Initializing _TorchClusterAreaMaskBackend on device cuda:0
2026-05-22 15:38:21 INFO Fitting AreaMaskBuilder (_TorchClusterAreaMaskBackend) with 822681 reference nodes from data[cutout_mask].
2026-05-22 15:38:21 DEBUG Reference nodes live on cuda:0
2026-05-22 15:38:24 INFO Computing area-mask (_TorchClusterAreaMaskBackend) for 2562 query nodes with margin: 11.0 km
2026-05-22 15:38:24 DEBUG Query nodes live on cpu
2026-05-22 15:38:24 DEBUG _TorchClusterAreaMaskBackend: Getting the query coordinate mask with Eucl. chord threshold 0.0017
2026-05-22 15:38:24 DEBUG _TorchClusterAreaMaskBackend: Moving query vector coordinates from cpu to cuda:0
2026-05-22 15:38:25 INFO Computing area-mask (_TorchClusterAreaMaskBackend) for 40962 query nodes with margin: 11.0 km
2026-05-22 15:38:25 DEBUG Query nodes live on cpu
2026-05-22 15:38:25 DEBUG _TorchClusterAreaMaskBackend: Getting the query coordinate mask with Eucl. chord threshold 0.0017
2026-05-22 15:38:25 DEBUG _TorchClusterAreaMaskBackend: Moving query vector coordinates from cpu to cuda:0
2026-05-22 15:38:25 DEBUG Time to register node coordinates (StretchedTriNodes): 5.22 s
2026-05-22 15:38:25 DEBUG Time to register node coordinates (StretchedTriNodes): 0.00 s
```

File: graphs/src/anemoi/graphs/generate/masks.py [hunk 2]
```diff
@@ -9,51 +9,167 @@
...
-    mask_attr_name : str
-        Name of a node to attribute to mask the reference nodes, if desired. Defaults to consider all reference nodes.
+class _TorchClusterAreaMaskBackend:
+    """Torch-cluster radius backend (CPU/GPU depending on distributed device)."""
+
+    def __init__(self, device: torch.device | str):
+        LOGGER.debug("Initializing %s on device %s", self.__class__.__name__, device)
+        self.device = device
+        self._ref_vectors: torch.Tensor | None = None
+        self._kdtree: cKDTree | None = None
```
Comment (mpvginde) [inline graphs/src/anemoi/graphs/generate/masks.py]: I think we can remove this `_kdtree` attribute here?

If yes, we could also move the `ckDTree` import to the `_KDTreeMaskBackend`, so we don't import when not needed.

File: graphs/tests/generate/test_mask_builder.py [hunk 3]
```diff
@@ -7,51 +7,115 @@
...
+
+def test_get_mask(graph_with_nodes: HeteroData):
+    """Test AreaMaskBuilder get_mask on query points that match references."""
+    mask_builder = AreaMaskBuilder("test_nodes", margin_radius_km=100)
+    mask_builder.fit(graph_with_nodes)
+
+    query_coords_rad = graph_with_nodes["test_nodes"].x.cpu().numpy()
+    mask = mask_builder.get_mask(query_coords_rad)
+
+    assert isinstance(mask, np.ndarray)
```
Comment (mpvginde) [inline graphs/tests/generate/test_mask_builder.py]: This is causing the failing test.
The `get_mask` function is always returning a torch.Tensor, even with the `_KDTreeAreaMaskBackend`


```suggestion
    assert isinstance(mask, torch.Tensor)
```

File: graphs/tests/generate/test_mask_builder.py [hunk 4]
```diff
@@ -7,51 +7,115 @@
...
+def test_get_mask(graph_with_nodes: HeteroData):
+    """Test AreaMaskBuilder get_mask on query points that match references."""
+    mask_builder = AreaMaskBuilder("test_nodes", margin_radius_km=100)
+    mask_builder.fit(graph_with_nodes)
+
+    query_coords_rad = graph_with_nodes["test_nodes"].x.cpu().numpy()
+    mask = mask_builder.get_mask(query_coords_rad)
+
+    assert isinstance(mask, torch.Tensor)
+    assert mask.dtype == np.bool_
```
Comment (mpvginde) [inline graphs/tests/generate/test_mask_builder.py]: Also dtype is `torch`


```suggestion
    assert mask.dtype == torch.bool
```

**Comment by JPXKQX** [comment]
[Integration tests](https://github.com/ecmwf/anemoi-core/actions/runs/26498836610)

**Review by JPXKQX — APPROVED** [review]
Thanks Michiel for the PR, it looks ready to be merged!

**Review by anaprietonem — APPROVED** [review]
